### DISCUSSION TITLE
V9. useTransition in react native FlatList

### DISCUSSION THREAD

**Discussion by willroyMyles** [pmndrs/react-spring#1340] [Support]
Is there anyway to implement useTransition with a flatlist in react native?

**Comment by willroyMyles** [comment]
or anyway to get the back an array similar to v8

**Comment by lukebrandonfarrell** [comment]
This is a great questions. +1 we can't use the new transitions format in FlatList anymore.

**Reply by joshuaellis** [reply]
If you're able to provide a repro I can look into, perhaps a small codesandbox, I can look into it ⭐

**Reply by lukebrandonfarrell** [reply]
@joshuaellis 

I mean I think this discussion is more to ask for guidance on V9, and it's compatibility with React Native, more than a bug. 

In V9 of react-spring the `const transitions = useTransition(..)` hook returns a function which can be used as such `transitions(...render)` which is incompatible with `FlatList` which requires a data property. In V8, `useTransition` would return an Array of items, which could be used with FlatList as such:

```
<FlatList
  date={transitions}
/> 
```

Does that make sense?

**Reply by joshuaellis** [reply]
It does make sense yes, in this case no `react-spring` as it currently is would not work with FlatList... Interesting problem to solve certainly!

**Comment by lukebrandonfarrell** [comment]
I have actually came up with a solution for this. You can do the following:

```
const array = []
transitions((style, item, t, i) => array.push({ style, item, t, i }))

<FlatList
   data={array}
   renderItem={({ item }) => ...}
/>
```

You can then write `item.style`, `item.item`, `item.t` etc to get the transition variables.

@joshuaellis Do you see any issues with this approach? Ideally we'd probably want to include some sort of helper in `@react-spring/native`, if you could advise me, I could PR?

**Reply by lukebrandonfarrell** [reply]
@willroyMyles

**Reply by joshuaellis** [reply]
Hm, there might be a better way... you could attach a `SpringRef` and then pass the Controllers.... They have the `springs` & their `items`, `t` is the `Controller`

**Reply by joshuaellis** [reply]
I think a helper or some sort to resolve this in `native` is certainly welcome. But my worry is the amount of rendering it would cause.

**Reply by lukebrandonfarrell** [reply]
Do you have a more elaborate example on the usage of `SpringRef` in this case ?

**Reply by joshuaellis** [reply]
Sure, sorry I was working at the time!

[`SpringRef`](https://react-spring.io/common/imperatives-and-refs#using-refs) returns an API to control the Controller(s) of the animation. It's parameter `.current` will return an array of `Controllers`. So in the case of a transition you'd get a Controller per item. To clarify, this isn't the same as `useRef`.

So (and I haven't tried this out yet) it could work something like this:

```js
const ref = useSpringRef()

useTransition(myData, {
  ref,
  ...styles
})

return (
  <FlatList
   data={ref.current}
   renderItem={({ springs, item }) => ...}
  />
)
```

This might not work though, i'm not totally sure...

**Reply by lukebrandonfarrell** [reply]
@joshuaellis Thanks for the detailed example! I'll be sure to test this out and post the findings on this discussion

**Reply by lukebrandonfarrell** [reply]
If that works, it might be worth just adding it to docs, on how to useTransition with RN flatlist

**Reply by lukebrandonfarrell** [reply]
@joshuaellis So I tested this out, but logging `ref.current` always gives me an empty array, i.e. reproduce using:

```
const ref = useSpringRef()

useTransition(["a", "b", "c"], {
  ref,
})

console.log(ref.current) = []
```

Any ideas?

**Reply by joshuaellis** [reply]
Oh, sorry, so this is on 9.2.0 latest beta, (there was a bug with attaching refs) so you either have to update to the beta or wait till this evening when I'll release stable.

**Reply by lukebrandonfarrell** [reply]
Great, at least it's just an update issue, I'll check it out and get back if the solution works

**Reply by lukebrandonfarrell** [reply]
@joshuaellis So I got that working, and now each flat list item receives a `springs` value with a bunch of properties. What do I need to pass into the `View` to control the animation? I don't see a `styles` property in this list. I tried `<AnimatedView style={springs}>` but that didn't work.

<img width="814" alt="Screenshot 2021-05-27 at 12 02 15" src="https://user-images.githubusercontent.com/18139277/119815419-68cc3880-bee3-11eb-813e-0b5a3e598a52.png">

**Reply by joshuaellis** [reply]
So this is from the `useTransition` hook:

```js
const renderTransitions: TransitionFn = render => (
    <>
      {transitions.map((t, i) => {
        const { springs } = changes.get(t) || t.ctrl
        const elem: any = render({ ...springs }, t.item, t, i)
        return elem && elem.type ? (
          <elem.type
            {...elem.props}
            key={is.str(t.key) || is.num(t.key) ? t.key : t.ctrl.id}
            ref={elem.ref}
          />
        ) : (
          elem
        )
      })}
    </>
  )
```

We just spread the springs into a a new object, so I suggest that could "just" work?

**Reply by lukebrandonfarrell** [reply]
You mean doing this should work? `<AnimatedView style={springs}>`

**Reply by lukebrandonfarrell** [reply]
@joshuaellis  ?

**Reply by joshuaellis** [reply]
Sorry @lukebrandonfarrell I missed this. 

In theory it should work. If you're able to make a small codebox repo, I can have a look at the solution myself too.

**Reply by bassasaurus** [reply]
I just stumbled along this thread while trying to implement with RN FlatList. Is there any progress, or a useable workaround?
